overlord/snapstate/refresh.go

```diff
@@ -21,17 +21,87 @@ package snapstate
 
 import (
 	"fmt"
+	"os"
+	"path/filepath"
 	"sort"
 	"strings"
 
 	"github.com/snapcore/snapd/cmd/snaplock"
+	"github.com/snapcore/snapd/dirs"
 	"github.com/snapcore/snapd/sandbox/cgroup"
 	"github.com/snapcore/snapd/snap"
 )
 
-var (
-	pidsCgroupDir = cgroup.ControllerPathV1("pids")
-)
+// pidsOfSnap returns the association of security tags to PIDs.
+//
+// NOTE: This function returns non-empty result only if refresh-app-awareness
+// is enabled.
+//
+// The return value is a snapshot of the pids of a given snap, grouped by
+// security tag. The result may be immediately stale as processes fork and
+// exit but it has the following guarantee.
+//
+// If the per-snap lock is held while computing the set, then the following
+// guarantee is true: If a security tag is not among the result then no such
+// tag can come into existence while the lock is held.
+//
+// This can be used to classify the activity of a given snap into activity
+// classes, based on the nature of the security tags encountered.
+//
+// TODO: move this to sandbox/cgroup later.
+func pidsOfSnap(snapInfo *snap.Info) (map[string][]int, error) {
+	// pidsByTag maps security tag to a list of pids.
+	pidsByTag := make(map[string][]int, len(snapInfo.Apps)+len(snapInfo.Hooks))
+
+	// TODO: de-duplicate this with interfaces.SecurityTagGlob
+	// securityTagGlob is a glob matching all the security tags of the snap.
+	securityTagGlob := snap.AppSecurityTag(snapInfo.InstanceName(), "*")
+	walkFunc := func(path string, fileInfo os.FileInfo, err error) error {
+		// We are only interested in files and don't want to analyze parent errors.
+		if err != nil || fileInfo.IsDir() {
+			return err
+		}
+		// We are only interested in cgroup.procs files.
+		if filepath.Base(path) != "cgroup.procs" {
+			println("not cgroup.procs: " + path)
+			return nil
+		}
+		// We are only interested in things matching our security tag glob.
+		parent, _ := filepath.Split(path)
+		securityTag := filepath.Base(parent)
+		if matched, err := filepath.Match(securityTagGlob, securityTag); err != nil || !matched {
+			return err
+		}
+		// Now that we know it is interesting parse the pids and put them into
+		// a bin of the exact security tag.
+		pids, err := cgroup.PidsInFile(path)
+		if err != nil {
+			return err
+		}
+		pidsByTag[securityTag] = append(pidsByTag[securityTag], pids...)
+		return nil
+	}
+
+	// TODO: Currently we walk the entire cgroup tree. We could be more precise
```

Comment:
Hmmm, between this and the above walks for mapping of pids to security labels, perhaps the walking of /proc/*/attr/current isn't as bad as I thought. ;) (Note, I only glanced at this to see the direction of the draft and did not perform a full review).

Reference: https://github.com/snapcore/snapd/pull/7722#issuecomment-550494007